### PR TITLE
Generate API docs for Toolpad Core

### PR DESCRIPTION
**Pull request by apedroferreira**
Generate API docs for Toolpad Core - first the proptypes must be generated with `pnpm proptypes` and then API docs built with `pnpm docs:build:api`.

Restructured the documentation a bit kind of just to be more like the other MUI libraries I was taking the scripts from and what these scripts expect, but it's all for discussion of course!

API links from component pages are broken - if this PR looks good I think we can fix it here in the MUI monorepo: https://github.com/mui/material-ui/blob/next/packages/markdown/prepareMarkdown.js

**API Page:** https://deploy-preview-3536--mui-toolpad-docs.netlify.app/toolpad/core/api/app-provider/

### CODE REVIEW

File: .npmrc [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Enforce pnpm
 
-engine-strict = true
-auto-install-peers = true
+engine-strict=true
+auto-install-peers=true
 strict-peer-dependencies=false
```
Comment (oliviertassinari) [inline .npmrc:5]: It seems to me that a space is supposed to be present from https://docs.npmjs.com/cli/v9/configuring-npm/npmrc.

Comment (apedroferreira) [reply .npmrc:5]: Oh ok I was changing these for consistency as I had seen different spacing in different places, will make all of them include the spacing then.

File: packages/toolpad-core/src/AppProvider/AppProvider.tsx [hunk 2]
```diff
@@ -135,7 +134,543 @@ AppProvider.propTypes /* remove-proptypes */ = {
    * [Theme](https://mui.com/material-ui/customization/theming/) used by the app.
    * @default baseTheme
    */
-  theme: PropTypes.any,
+  theme: PropTypes.shape({
```
Comment (apedroferreira) [inline packages/toolpad-core/src/AppProvider/AppProvider.tsx:137]: wow

**Comment by apedroferreira** [comment]
TODO:
- [ ] Fix proptypes generation
- [ ] Do not list demos in callout for every component with API links

File: docs/data/toolpad/core/pages.ts [hunk 3]
```diff
@@ -60,86 +60,44 @@ const pages: MuiPage[] = [
...
-          {
-            pathname: '/toolpad/core/components/line-chart',
-            title: 'Line Chart',
+            pathname: '/toolpad/core/react-app-provider',
+            title: 'App Provider',
           },
         ],
       },
       {
-        pathname: '/toolpad/core/components/filter-group',
```
Comment (apedroferreira) [inline docs/data/toolpad/core/pages.ts:89]: Removing these but we can restore them back in the same way once we add more components, they might have been causing issues with the scripts.

File: docs/data/toolpad/core/pages.ts [hunk 4]
```diff
@@ -60,86 +60,44 @@ const pages: MuiPage[] = [
...
-        subheader: 'Filters',
+        pathname: '/toolpad/core/layout-group',
+        subheader: 'Layout',
         children: [
           {
-            pathname: '/toolpad/core/components/select-filter',
-            title: 'Select',
+            pathname: '/toolpad/core/react-dashboard-layout',
```
Comment (apedroferreira) [inline docs/data/toolpad/core/pages.ts:79]: Not sure if we should follow this `react-...` naming pattern for pages, I saw all the other MUI libraries are doing it? But we can do it differently too.

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx [hunk 5]
```diff
@@ -49,6 +48,60 @@ const LogoContainer = styled('div')({
   },
 });
 
+function ToolpadLogo({ size = 40 }: { size?: number }) {
```
Comment (apedroferreira) [inline packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:51]: Moved this component inside here as that's kind of how the scripts expect files to be organized, but it does take up a lot of lines...

**Review by Janpot — APPROVED** [review]

**Comment by Janpot** [comment]
Thanks! I removed an alias as it broke the demo formatter.

**Pull request merged**